pyface/ui/qt4/font.py

```diff
@@ -0,0 +1,202 @@
+# license included in LICENSE.txt and may be redistributed only under
+# the conditions described in the aforementioned license. The license
+# is also available online at http://www.enthought.com/licenses/BSD.txt
+#
+# Thanks for using Enthought open source!
+
+"""
+Font conversion utilities
+
+This module provides facilities for converting between pyface Font objects
+and Qt QFont objects, trying to keep as much similarity as possible between
+them.
+"""
+
+from pyface.qt import is_qt4
+from pyface.qt.QtGui import QFont
+
+
+qt_family_to_generic_family = {
+    QFont.AnyStyle: 'default',
+    QFont.System: 'default',
+    QFont.Decorative: 'fantasy',
+    QFont.Serif: 'serif',
+    QFont.Cursive: 'cursive',
+    QFont.SansSerif: 'sans-serif',
+    QFont.Monospace: 'monospace',
+    QFont.TypeWriter: 'typewriter',
+}
+generic_family_to_qt_family = {
+    'default': QFont.System,
+    'fantasy': QFont.Decorative,
+    'decorative': QFont.Decorative,
+    'serif': QFont.Serif,
+    'roman': QFont.Serif,
+    'cursive': QFont.Cursive,
+    'script': QFont.Cursive,
+    'sans-serif': QFont.SansSerif,
+    'swiss': QFont.SansSerif,
+    'monospace': QFont.Monospace,
+    'modern': QFont.Monospace,
+    'typewriter': QFont.TypeWriter,
+}
+
+weight_to_qt_weight = {
+    100: 0,
+    200: QFont.ExtraLight,
+    300: QFont.Light,
+    400: QFont.Normal,
+    500: QFont.Medium,
+    600: QFont.DemiBold,
+    700: QFont.Bold,
+    800: QFont.ExtraBold,
+    900: QFont.Black,
+    1000: 99,
+}
+qt_weight_to_weight = {
+    0: 'thin',
+    QFont.ExtraLight: 'extra-light',
+    QFont.Light: 'light',
+    QFont.Normal: 'normal',
+    QFont.Medium: 'medium',
+    QFont.DemiBold: 'demibold',
+    QFont.Bold: 'bold',
+    QFont.ExtraBold: 'extra-bold',
+    QFont.Black: 'black',
+    99: 'extra-heavy',
+}
+
+style_to_qt_style = {
+    'normal': QFont.StyleNormal,
+    'oblique': QFont.StyleOblique,
+    'italic': QFont.StyleItalic,
+}
+qt_style_to_style = {value: key for key, value in style_to_qt_style.items()}
+
+
+def font_to_toolkit_font(font):
+    """ Convert a Pyface font to a Qfont.
+
+    Parameters
+    ----------
+    font : pyface.font.Font
+        The Pyface font to convert.
+
+    Returns
+    -------
+    qt_font : QFont
+        The best matching Qt font.
+    """
+    qt_font = QFont()
+    families = []
+    default_family = None
+
+    for family in font.family:
+        if family not in generic_family_to_qt_family:
+            families.append(family)
+        elif default_family is None:
+            default_family = family
+
+    if families and hasattr(qt_font, 'setFamilies'):
+        # Qt 5.13 and later
+        qt_font.setFamilies(families)
+    elif families:
+        qt_font.setFamily(families[0])
+        # Note: possibily could use substitutions here,
+        # but not sure if global (which would be bad, so we don't)
+
+    if default_family is not None:
+        qt_font.setStyleHint(generic_family_to_qt_family[default_family])
```

Comment:
We're using `default_family` to set the style hint but we're not using using it for `setFamily`. I'm not sure why.

Reply:
The way that Qt font selection works (prior to 5.13, and ignoring font substitutions, which I think are global state so we don't want to mess with them), is that you specify the font that you want (which we are taking as the _first_ font specified) via `setFamily` and then the fall-back _generic_ font type via `setStyleHint` (so we look for one of the special matching generic names, which are a union of Qt, Wx, and CSS special names).